### ISSUE TITLE
[HOLD for payment 2024-08-09] [$250]  ‘New messages’ banner shows up every time for the sender when sending a message to a chat report 

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#43486]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.81-8
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @youssef-lr 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1718058418632669

## Action Performed:
1. Go to a chat report (user B) with lots of messages as user A
2. Send a message from user A to user B

## Expected Result:
`New message` banner should not appear for the sender (user A)

## Actual Result:
 ‘New messages’ banner shows up every time sending a new message in a chat, the message is coming from user A
## Workaround:
unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [x] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/38435837/03d2d7eb-034a-425d-8bd3-482c5a57a60d

https://github.com/Expensify/App/assets/38435837/01e377c7-bce7-4a16-9442-271c4287a556

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e44bca9dc779dfff</li>
        <li>Upwork Job ID: 1803494636295070981</li>
        <li>Last Price Increase: 2024-06-26</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Reviewer | 102953243</li>
<li>bernhardoj | Contributor | 102953244</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @alexpensify</details>

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
New message floating banner shows every time adding a new message.

### What is the root cause of that problem?
One of the condition that will show the new message floating banner is `canScrollToNewerComments`. In our case, when sending a new message, `hasNewestReportAction` becomes false, so `canScrollToNewerComments` becomes true.
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/pages/home/report/ReportActionsList.tsx#L679
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/pages/home/report/ReportActionsList.tsx#L682-L683

`hasNewestReportAction` checks whether we have the newest report action or not.
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/pages/home/report/ReportActionsList.tsx#L215

When we send a new message, we optimistically update both report and report actions.
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/libs/actions/Report.ts#L511-L522

However, the component received the report optimistic update first which makes `hasNewestReportAction` becomes false. Later, when the component receive the report actions optimistic update, `hasNewestReportAction` becomes true again. _Both report and report actions are received from ReportScreen._

It's because for report onyx, we use useOnyx,
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/pages/home/ReportScreen.tsx#L151

and for report actions, we use withOnyx,
https://github.com/Expensify/App/blob/95c8202a0e8fdf133fc65bd8d249c7d91372e2ff/src/pages/home/ReportScreen.tsx#L760-L764

so their state update are not batched together.

### What changes do you think we should make in order to solve the problem?

Convert report actions from withOnyx to useOnyx.

```
const [sortedAllReportActions] = useOnyx(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${getReportID(route)}`, {canEvict: false,
        selector: (allReportActions: OnyxEntry<OnyxTypes.ReportActions>) => ReportActionsUtils.getSortedReportActionsForDisplay(allReportActions, true)})
```

_Maybe we can convert all withOnyx to useOnyx, but withOnyx has shouldDelayUpdate set to true for ReportScreen which isn't present in useOnyx, so not sure._

**Comment by alexpensify** [comment]
I'll test this one soon.

**Comment by alexpensify** [comment]
No update yet

**Comment by alexpensify** [comment]
@youssef-lr - I followed up in the Slack thread since I'm unable to replicate the experience: 

![2024-06-18_12-08-29](https://github.com/Expensify/App/assets/8107070/7fcafadb-4355-4237-935f-861cdfac1dc0)

**Comment by alexpensify** [comment]
I was able to replicate the experience if you overload a chat with a ton of activity. Also, I think the issue might be the response trying to catch up. I'm going to mark this as external, but I think @youssef-lr should be involved in the review.  

https://github.com/Expensify/App/assets/8107070/d6bc91bc-d59c-4390-a439-7d4805f58342

**Comment by alexpensify** [comment]
@shubham1206agra can you please review if the proposal here will fix the issue? Thanks!

**Comment by alexpensify** [comment]
@shubham1206agra any update regarding the proposal here? Thanks!

**Comment by bernhardoj** [comment]
@shubham1206agra here is the [test branch](https://github.com/bernhardoj/Expensify/tree/fix/43486-new-message-shows-every-time-add-new-message)

**Comment by youssef-lr** [comment]
>Maybe we can convert all withOnyx to useOnyx, but withOnyx has shouldDelayUpdate set to true for ReportScreen which isn't present in useOnyx, so not sure.

@bernhardoj `withOnyx` is now deprecated, we can safely use `useOnyx` here

**Comment by alexpensify** [comment]
@bernhardoj - can you please update your proposal based on the feedback here? Thanks!

**Comment by bernhardoj** [comment]
Updated my [proposal](https://github.com/Expensify/App/issues/43486#issuecomment-2162491991)

**Comment by alexpensify** [comment]
@youssef-lr - assigning you since we need your help to test the branch. Thanks!

**Comment by youssef-lr** [comment]
Sounds good @alexpensify, I don't think I can get to it today, will do tomorrow. I can point it to staging and see if it fixes the issue.

**Comment by alexpensify** [comment]
Sounds like a plan!

**Comment by alexpensify** [comment]
Waiting for a branch test

**Comment by alexpensify** [comment]
@youssef-lr any update to test here? I'm trying to keep moving this one forwards. Thanks! 

--- 

## Heads up, I will be offline until Monday, July 8, 2024, and will not actively watch over this GitHub during that period.

If this GitHub needs an urgent update, please ask for help in the #expensify-open-source Slack Room. If it can wait, I'll continue the review process when I return online. Thanks!

**Comment by youssef-lr** [comment]
Yeah it tested well for me in Chronos, but haven't tested it in other messages, will do tonight. But looking good!

**Comment by youssef-lr** [comment]
@bernhardoj please proceed with the PR!

**Comment by bernhardoj** [comment]
PR is ready

cc: @shubham1206agra

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by alexpensify** [comment]
Catching up from being OOO, this PR is under review.

**Comment by alexpensify** [comment]
PR is still under review but moving along.

**Comment by bernhardoj** [comment]
We have 2 issues found after the PR which causes it to be reverted.

1. https://github.com/Expensify/App/issues/45657
This issue is ultimately caused by https://github.com/Expensify/App/pull/44955 where we replace `useLastAccessedReportID` hook and instead use `ReportUtils.findLastAccessedReport`. 
```
const lastAccessedReportID = useLastAccessedReportID(!!route.params.openOnAdminRoom);
to
const lastAccessedReportID = ReportUtils.findLastAccessedReport(!canUseDefaultRooms, !!route.params.openOnAdminRoom, activeWorkspaceID)?.reportID;
```
It's only uncovered by our PR where we remove the `markForHydration` usage.

When we still use `markForHydration`, react-compiler can't compile ReportScreen. As soon as we remove it, the react-compiler works for ReportScreen.

https://github.com/user-attachments/assets/c52df5b8-97f2-43b2-a2df-5d7e36363985

This means it will memoize value automatically including `findLastAccessedReport`. It will recalculate only if the function params (canUseDefaultRooms, route.params.openOnAdminRoom, activeWorkspaceID) are changed. `findLastAccessedReport` depends on "reports" inside the function, but because we don't recalculate the function when reports are changed, the value returned is always undefined.

We have this PR https://github.com/Expensify/App/pull/45494 that is the follow-up of https://github.com/Expensify/App/pull/44955 where we will call `findLastAccessedReport` inside a `useEffect` that is triggered when the app loading is completed. So far from my testing, when the loading completes, the "reports" data are ready. I think it would be more future-proof if we check for the "reports" availability, but I don't think we want to subscribe to it in ReportScreen as that means it will re-render every time any report is updated.

So, the solution for this is to hold for https://github.com/Expensify/App/pull/45494

2. https://github.com/Expensify/App/issues/45634
I'm still investigating this one.

**Comment by bernhardoj** [comment]
Opened a new PR.

cc: @shubham1206agra

**Comment by alexpensify** [comment]
Noted the payment date is tomorrow

**Comment by alexpensify** [comment]
@shubham1206agra, thank you for that update! I've updated the GH. We will continue with the payment process after this second PR goes into production and then clears the regression period.

**Comment by alexpensify** [comment]
Weekly Update: Waiting on https://github.com/Expensify/App/pull/45711

**Comment by alexpensify** [comment]
I was OOO on Friday and will work on the payment process tomorrow.

**Comment by alexpensify** [comment]
Payouts due: 2024-08-09

- [ ] Contributor: $250 @bernhardoj 
- [ ] Contributor+: $250 @shubham1206agra 

Upwork job is [here](https://www.upwork.com/jobs/~01e44bca9dc779dfff).

**Comment by bernhardoj** [comment]
Requested in ND.

**Comment by alexpensify** [comment]
I've completed the payment process in Upwork - https://github.com/Expensify/App/issues/43486#issuecomment-2284765265

Closing for now since the remaining payment will be handled via Chat.

**Comment by JmillsExpensify** [comment]
$250 approved for @bernhardoj
